File: Problems/0143. (M) Reorder List/Python3/solution.py
class Solution:
  def reorderList(self, head: Optional[ListNode]) -> None:
    # store all nodes in queue
    # then first and last node pair is i-th and (n-i)-th nodes

    nodes = deque()
    cur = head
    while cur:
      nodes.append(cur)
      cur = cur.next
    
    prev_tail = head
    while len(nodes) > 1:
      left, right = nodes.popleft(), nodes.pop()
      prev_tail.next = left
      left.next = right
      prev_tail = right
    
    # if list has odd number of nodes
    if nodes:
      prev_tail.next = nodes.pop()
      prev_tail = prev_tail.next
    # tail node should point to None
    prev_tail.next = None

    # head is unchanged
    return head

File: Problems/0143. (M) Reorder List/Java/solution.java
class Solution {
  public void reorderList(ListNode head) {
    // store all nodes in queue
    // then first and last nodes are i-th and (n-i)-th nodes

    // removeFirst/Last are Deque interfaces
    Deque<ListNode> nodes = new ArrayDeque();
    ListNode cur = head;
    while (cur != null) {
      nodes.add(cur);
      cur = cur.next;
    }

    ListNode prev_tail = head;
    while (nodes.size() > 1) {
      ListNode left = nodes.removeFirst();
      ListNode right = nodes.removeLast();
      prev_tail.next = left;
      left.next = right;
      prev_tail = right;
    }

    // if list has odd number of nodes
    if (nodes.size() == 1) {
      prev_tail.next = nodes.remove();
      prev_tail = prev_tail.next;
    }
    // tail node should point to null
    prev_tail.next = null;
  }
}
